kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/client/DefaultJsonHandlerSuite.scala

```diff
@@ -0,0 +1,110 @@
+
+  //////////////////////////////////////////////////////////////////////////////////
+  // END-TO-END TESTS FOR parseStructType (more tests in DataTypeParserSuite)
+  //////////////////////////////////////////////////////////////////////////////////
+
+  private def sampleMetadata: FieldMetadata = FieldMetadata.builder()
+    .putNull("null")
+    .putLong("long", 1000L)
+    .putDouble("double", 2.222)
+    .putBoolean("boolean", true)
+    .putString("string", "value")
+    .build()
+
+  test("parseStructType: primitive type round trip") {
+    val fields = BasePrimitiveType.getAllPrimitiveTypes().asScala.flatMap { dataType =>
+      Seq(
+        new StructField("col1" + dataType, dataType, true),
+        new StructField("col1" + dataType, dataType, false),
+        new StructField("col1" + dataType, dataType, false, sampleMetadata)
+      )
+    } ++ Seq(
+      new StructField("col1decimal", new DecimalType(30, 10), true),
+      new StructField("col2decimal", new DecimalType(38, 22), true),
+      new StructField("col3decimal", new DecimalType(5, 2), true)
+    )
+
+    val expSchema = new StructType(fields.asJava);
+    val serializedSchema = expSchema.toJson
+    val actSchema = jsonHandler.parseStructType(serializedSchema)
+    assert(expSchema == actSchema)
+  }
+
+  test("parseStructType: complex type round trip") {
+    val arrayType = new ArrayType(IntegerType.INTEGER, true)
+    val arrayArrayType = new ArrayType(arrayType, false)
+    val mapType = new MapType(FloatType.FLOAT, BinaryType.BINARY, false)
+    val mapMapType = new MapType(mapType, BinaryType.BINARY, true)
+    val structType = new StructType().add("simple", DateType.DATE)
+    val structAllType = new StructType()
+      .add("prim", BooleanType.BOOLEAN)
+      .add("arr", arrayType)
+      .add("map", mapType)
+      .add("struct", structType)
+
+    val expSchema = new StructType()
+      .add("col1", arrayType, true)
+      .add("col2", arrayArrayType, false)
+      .add("col3", mapType, false)
+      .add("col4", mapMapType, false)
+      .add("col5", structType, false)
+      .add("col6", structAllType, false)
+
+    val serializedSchema = expSchema.toJson
+    val actSchema = jsonHandler.parseStructType(serializedSchema)
+    assert(expSchema == actSchema)
+  }
+
+  test("parseStructType: not a StructType") {
+    val e = intercept[IllegalArgumentException] {
+      jsonHandler.parseStructType(new ArrayType(StringType.STRING, true).toJson())
+    }
+    assert(e.getMessage.contains("as a valid StructType"))
+  }
+
+  test("parseStructType: invalid JSON") {
+    val e = intercept[RuntimeException] {
+      jsonHandler.parseStructType(
+        """
+          |{
+          |  "type" : "struct,
+          |  "fields" : []
+          |}
+          |""".stripMargin
+      )
+    }
+    assert(e.getMessage.contains("Could not parse JSON"))
+  }
+
+  // TODO we use toJson to serialize our physical and logical schemas in ScanStateRow, we should
```

Comment:
Looks like you have migrated the tests from the deleted suite `TestTableSchemaSerDe` here which uses `DataType.toJson`. Do we need more tests? If not can we remove the TODO here?

Reply:
I think the two tests aren't really enough. But that code is all in kernelAPI still and will be replaced later so not completely sure what we should do here.

Reply:
Got it. Could you please create issues for this TODO and the other TODO about FieldMetadata.

Reply:
https://github.com/delta-io/delta/issues/2313 https://github.com/delta-io/delta/issues/2315